File: leetcode/binary-tree-level-order-traversal.java
//https://leetcode.com/problems/binary-tree-level-order-traversal/

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class Solution {

    //Provided TreeNode class
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ordered = new ArrayList<>();
        if(root == null) return ordered;
        
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        
        while(queue.size() > 0){
            List<Integer> level = new ArrayList<Integer>();
            int size = queue.size();
            
            for(int i = 0; i<size; i++){
                TreeNode curr = queue.poll();
                level.add(curr.val);
                if(curr.left != null) queue.add(curr.left);
                if(curr.right != null) queue.add(curr.right);
            }
            ordered.add(level);
        }
        
        return ordered;        
    }
}

File: leetcode/binary_tree_level_order_traversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal/description

from typing import Optional

# Provided TreeNode class
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> list[list[int]]:
        order: list[list[int]] = []
        queue: list[int] = []

        queue.append(root)

        while len(queue) != 0:
            level: list[int] = []
            next_level: list[int] = []
            while len(queue) != 0:
                node: Optional[TreeNode] = queue.pop()
                if node is not None:
                    level.append(node.val)
                    if node.left is not None:
                        next_level.insert(0, node.left)
                    if node.right is not None:
                        next_level.insert(0, node.right)
            if len(level) > 0:
                order.append(level)
            queue.extend(next_level)

        return order
